pkg/collector/reconcile/statefulset.go

```diff
@@ -0,0 +1,135 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconcile
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/open-telemetry/opentelemetry-operator/pkg/collector"
+)
+
+// +kubebuilder:rbac:groups="apps",resources=statefulsets,verbs=get;list;watch;create;update;patch;delete
+
+// StatefulSets reconciles the stateful set(s) required for the instance in the current context.
+func StatefulSets(ctx context.Context, params Params) error {
+
+	desired := []appsv1.StatefulSet{}
+	if params.Instance.Spec.Mode == "statefulset" {
+		desired = append(desired, collector.StatefulSet(params.Config, params.Log, params.Instance))
+	}
+
+	// first, handle the create/update parts
+	if err := expectedStatefulSets(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the expected stateful sets: %v", err)
+	}
+
+	// then, delete the extra objects
+	if err := deleteStatefulSets(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the stateful sets to be deleted: %v", err)
+	}
+
+	return nil
+}
+
+func expectedStatefulSets(ctx context.Context, params Params, expected []appsv1.StatefulSet) error {
+	for _, obj := range expected {
+		desired := obj
+
+		if err := controllerutil.SetControllerReference(&params.Instance, &desired, params.Scheme); err != nil {
+			return fmt.Errorf("failed to set controller reference: %w", err)
+		}
+
+		existing := &appsv1.StatefulSet{}
+		nns := types.NamespacedName{Namespace: desired.Namespace, Name: desired.Name}
+		err := params.Client.Get(ctx, nns, existing)
+		if err != nil && k8serrors.IsNotFound(err) {
+			if err := params.Client.Create(ctx, &desired); err != nil {
+				return fmt.Errorf("failed to create: %w", err)
+			}
+			params.Log.V(2).Info("created", "statefulset.name", desired.Name, "statefulset.namespace", desired.Namespace)
+			continue
+		} else if err != nil {
+			return fmt.Errorf("failed to get: %w", err)
+		}
+
+		// it exists already, merge the two if the end result isn't identical to the existing one
+		updated := existing.DeepCopy()
+		if updated.Annotations == nil {
+			updated.Annotations = map[string]string{}
+		}
+		if updated.Labels == nil {
+			updated.Labels = map[string]string{}
+		}
+
+		updated.Spec = desired.Spec
```

Comment:
Could you double-check that you can copy the entire Spec here? Some resources have properties that can't be changed, or that we don't want to change.

Reply:
I have double checked [StatefulSetSpec](https://github.com/kubernetes/api/blob/3a2d6b5bb7f84daa2125131865eb81b151448d5b/apps/v1beta1/types.go#L162), and didn't see any fields that we do not want to change. Any ideas?